File: discipline_ts/core/rules/time_range_rules.ts
import { TimeRangeRule, Nominal } from "../x.ts";

const BRAND = Symbol();

export type TimeRangeRules = Nominal<typeof BRAND, TimeRangeRule[]>;

export const construct = (rules: TimeRangeRule[]): TimeRangeRules => {
  return Nominal.create(BRAND, rules);
};

export const createDefault = (): TimeRangeRules => {
  return construct([]);
};

export const TimeRangeRules = {
  createDefault,
};

File: discipline4/app/src/main/java/com/example/app/rules/TimeRangeRules.kt
package com.example.app

import com.example.app.*

public data class TimeRangeRules(
  val rules: Map<String, TimeRangeRule>,
) {
  companion object {
    fun createDefault(): TimeRangeRules {
      return TimeRangeRules(rules = emptyMap())
    }
  }

  fun isActiveAt(instant: Instant, time: Time): Boolean {
    return rules.values.any { rule -> rule.isActiveAt(instant, time) }
  }
}